cmd/snap/cmd_desktop_launch.go

```diff
@@ -0,0 +1,111 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/desktop/desktopentry"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/i18n"
+)
+
+type cmdDesktopLaunch struct {
+	waitMixin
+
+	DesktopFile string `long:"desktop" required:"true"`
+	Action      string `long:"action"`
+	Positional  struct {
+		FilesOrUris []string `positional-arg-name:"<files-or-uris>" required:"0"`
+	} `positional-args:"true"`
+}
+
+func init() {
+	addRoutineCommand("desktop-launch",
+		i18n.G("Launch a snap application via a desktop file"),
+		i18n.G("The desktop-launch command is a helper used to launch desktop entries."),
+		func() flags.Commander {
+			return &cmdDesktopLaunch{}
+		}, nil, nil)
+}
+
+func cmdlineArgsToUris(args []string) ([]string, error) {
+	uris := make([]string, len(args))
+	for i, arg := range args {
+		// This follows the logic of glib's
+		// g_file_new_for_commandline_arg function:
+		//  - if it looks like an absolute path, it is a file path
+		//  - if it has a valid URI scheme, it is a URI
+		//  - otherwise, it is a relative file path
+		if filepath.IsAbs(arg) {
+			u := &url.URL{Scheme: "file", Path: arg}
+			uris[i] = u.String()
+		} else if u, err := url.Parse(arg); err == nil && u.IsAbs() {
+			uris[i] = arg
+		} else {
+			fullPath, err := filepath.Abs(arg)
+			if err != nil {
+				return nil, err
```

Comment:
do we need a test for this case? also should the error be prefixed to clarify what we are trying to do?

Reply:
From the look of it, this can only fail if `os.Getwd()` fails. That in turn calls the `getcwd` system call. I'm not sure if there is an easy way to trigger this to fail.